### PR TITLE
fix: premature close with chunked transfer encoding and for async iterators in Node 12

### PR DESCRIPTION
**Pull request by achingbrain**
This PR backports the fix from #1064 to the `2.x.x` branch following the [comment here](https://github.com/node-fetch/node-fetch/pull/1064#issuecomment-849167400).

I had to add some extra babel config to allow using the `for await..of` syntax in the tests.  The config is only needed for the tests as this syntax is not used in the implementation.

### CODE REVIEW

File: test/test.js
Comment (achingbrain) [inline test/test.js:1089]: This is a bit weird - the error event gets emitted on the response body twice in node 14 (and 12 with this PR) - once with the expected `AbortError` and then again with `ERR_STREAM_PREMATURE_CLOSE`.

Both errors were still emitted even after changing how the `AbortError` is emitted (e.g. using `.destroy(err)` instead of `.emit('error', err)` so I'm open to suggestions on how to not need to change `.on` to `.once` here, if it's an issue.

**Comment by achingbrain** [comment]
Ok, I hadn't realised quite how far back in time the supported node versions would go, will see if I can fix up some the oldies.

File: src/index.js
Comment (achingbrain) [inline src/index.js:314]: In #1064 this was `socket.prependListener` but that's not supported in node 4.

File: src/index.js
Comment (achingbrain) [inline src/index.js:116]: In #1064 this was `s.prependListener` but that's not supported in node 4.

File: src/index.js
Comment (achingbrain) [inline src/index.js:112]: In #1064 this was `if (process.version < 'v14')` but it fails for node 8.

File: src/index.js
Comment (achingbrain) [inline src/index.js:118]: #1064 used the count of types of events with listeners registered on the socket to determine if the stream was closed prematurely but it's not reliable on node 8 and below, so instead if we've got a listener for `'data'` events we're still expecting data to arrive so use that to work out if we should emit the `ERR_STREAM_PREMATURE_CLOSE` error.

**Comment by achingbrain** [comment]
I had to pin a few deps that have dropped compatibility with old node versions without releasing majors and I've added notes where this PR diverges from the fix in #1064 for the sake of compatibility with old node versions but this should be good for review.

File: test/test.js
Comment (gr2m) [reply test/test.js:1092]: not sure about this, let's see what they other maintainers think, otherwise A++ pull request, thanks!

**Review by gr2m — APPROVED** [review]

File: test/test.js
Comment (achingbrain) [reply test/test.js]: I've refactored this away - now the implementation will check to see if the request has been aborted before emitting the `ERR_STREAM_PREMATURE_CLOSE` errors.

CI also now tests on all stable node versions 4-16 for a bit more confidence.

**Review by gr2m — APPROVED** [review]
thanks, looks good now!

I do not have access to the npm package myself, so someone else from @node-fetch/core needs to merge and publish the new version

File: src/index.js
Comment (achingbrain) [inline src/index.js:313]: In #1064 this checks the incoming bytes to make sure it ends with `0\r\n` which signifies the last chunk is received.  I found this didn't work well against real-world servers as it assumes the `'data'` event would align with incoming chunks whereas I don't think there's any such guarantee.

I've reused the same check as the [patch for node 10-12 above](https://github.com/node-fetch/node-fetch/pull/1172/files#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556R122) which seems to work well.

**Comment by oed** [comment]
Hey @xxczaki @bitinn @TimothyGu! Can one of you please have a look an release this? My team is facing huge issues that should be fixed with this patch backported to 2.x.x. Should only take a few minutes of your time!

File: src/index.js
Comment (tekwiz) [reply src/index.js:312]: @achingbrain The data event doesn't have to align; this check is only applicable at the very end when the socket is closed and the last bytes of data are pushed out to the data event.  The only way for a chunked transfer to indicate "the end" is for `0\r\n` to be the last 3 bytes at this point, which is what the check was for.

**Comment by glcohen** [comment]
bumping, @jimmywarting @achingbrain can this be merged?

**Review by tekwiz — APPROVED** [review]

**Comment by tekwiz** [comment]
@jimmywarting I'm good with merging this into 2.x, but I feel it's significant enough to warrant a minor version bump to 2.7.0. I'll leave it to you to merge so you can make the determination on versioning.

**Comment by IvanDeluxe** [comment]
Hey :)

Is there a way I could use this even though it's not merged .. or can we just merge it? 🔥

**Comment by aphelionz** [comment]
friendly bump :)

**Comment by achingbrain** [comment]
Thanks so much for merging this. I don't think it's been released yet, is there any chance of getting it out of the door?

**Comment by gr2m** [comment]
@jimmywarting we don't have semantic-release setup in the `2.x` branch, we need to release manually, or setup semantic-release in there as well. Let me know if you'd like help with the latter

**Comment by jimmywarting** [comment]
@gr2m if you could set it up, then it would be great

**Comment by achingbrain** [comment]
@gr2m @jimmywarting just a gentle reminder this still hasn't made it into a release.  Please could the 2.x.x branch have a release with this in it?

**Comment by achingbrain** [comment]
@gr2m @jimmywarting please, please, please can this be released?

**Comment by gr2m** [comment]
hey sorry I'll look into it

**Comment by gr2m** [comment]
there y'all go. Can you confirm it looks good?

**Comment by gr2m** [comment]
@node-fetch/core I've setup automated releases in the 2.x branch, so backporting should be easy now, just send PRs against the 2.x branch and use [the commit conventions](https://github.com/node-fetch/node-fetch/blob/main/CONTRIBUTING.md#maintainers-only)

**Comment by achingbrain** [comment]
Thankyouthankyouthankyouthankyou!

File: src/index.js
Comment (steveluscher) [reply src/index.js:312]: This has caused problems when reusing sockets for multiple requests, as when you use an `https.Agent` in ‘keep-alive’ mode.

https://github.com/node-fetch/node-fetch/issues/1767

**Pull request merged**
